R/slice.R

```diff
@@ -18,14 +22,18 @@
 #' x
 #'
 #' vec_slice(mtcars, 1:3)
-vec_slice <- function(x, i) {
-  .Call(vctrs_slice, x, maybe_missing(i))
+vec_slice <- function(x, i, ...) {
+  return(.Call(vctrs_slice, x, maybe_missing(i), environment()))
```

Comment:
Actually, are you sure we need `...` here?

Reply:
This way we can use `vec_slice()` in the `[` methods of composite structures like data frames and records, and ensure the `...` are properly passed to `[` from `vec_slice.default()` on recursion.

Reply:
Something seems faulty about that reasoning but I can't figure out what. Maybe it's that calling `[` is just a mostly a stop gap until we have our own implementations? Maybe that I see `vec_slice()` as considerably more restrictive than `[` so we shouldn't accept `...` without deliberation.

Reply:
Calling `[` from `vec_slice.default()` seems needed to be compatible with foreign types. And calling `vec_slice()` from `[.vctrs_xxx` is needed to get strict semantics with the subset syntax for our own types. I agree the proposed dispatch path is a bit convoluted but it seems like a reasonable compromise.

Would you prefer to call `check_dots_empty()` from `[.vctrs_xxx`?

Reply:
Yeah, but why would they need extra arguments? `vec_slice()` is more restrictive.

Reply:
For the recursive case so we can dispatch to `[` from `vec_slice.default` and forward the dots passed by the user to `[.vctrs_xxx`.

Reply:
Here is the dispatch path for a record that contains a vector of class `foo`:

1. `[.vctrs_rcrd()`
2. `vec_slice.vctrs_rcrd()`
3. For each element: `vec_slice()`
4. Because `foo` doesn't implement vec_slice: `vec_slice.default()`
5. `[.foo`

The `...` are forwarded through all these levels.

Reply:
Yes, but I still don't understand why we need to forward `...` — `[.vctrs_rcrd` implements the `vec_slice()` interface not the more general `[` interface.

Reply:
Because we use `[` has a fallback for compatibility with foreign classes. And we have recursive types like records and data frames which may contain any class of vector.

The alternative is to check for empty dots in `[.vctrs_xxx` (this would be a change from current master which forwards dots).